appsec/src/helper/sampler.hpp

```diff
@@ -6,42 +6,285 @@
+        auto cur_data = entry.data.load(std::memory_order_relaxed);
+        if (cur_data.last_reported <= report_threshold) {
+            // potentially a hit
+            typename entry::edata desired_data{now, now};
+
+        retry:
+            if (!entry.data.compare_exchange_strong(
+                    cur_data, desired_data, std::memory_order_relaxed)) {
+                // another thread just updated it
+                // was it a hit?
+                if (cur_data.last_accessed == cur_data.last_reported) {
+                    // then this one should not be a hit
+                    return false;
+                }
+
+                // the other thread did not register a hit
+                // we retry if our idea of time is ahead
+                if (cur_data.last_accessed < now) {
+                    goto retry;
+                }
+            }
+
             return true;
+        } else {
+            // we just update the last accessed time
+            typename entry::edata desired_data{now, cur_data.last_reported};
+
+            // this would only be slightly innacurate:
+            // entry.data.store(desired_data, std::memory_order_relaxed);
+
+        retry2:
+            if (cur_data.last_accessed >= now) {
+                // we're behind the times
+                return false;
+            }
+
+            if (!entry.data.compare_exchange_strong(
+                    cur_data, desired_data, std::memory_order_relaxed)) {
+                goto retry2;
+            }
+
+            return false;
+        }
+    }
+
+public:
+    class table_accessor : private rcu_reader_state<table> {
+    public:
+        table_accessor(timed_set &set)
+            : rcu_reader_state<table>{set.rcu}, set{set}
+        {}
+
+        bool hit(std::uint64_t number) noexcept
+        {
+            auto guard{this->lock()};
+            auto &table = *guard;
+            return set.hit(table, number);
+        }
+
+        // for testing
+        auto approx_size() noexcept
+        {
+            auto guard{this->lock()};
+            return guard->size.load(std::memory_order_relaxed);
+        }
+
+    private:
+        timed_set &set;
+    };
+    friend class table_accessor;
+
+    ~timed_set()
+    {
+        // we must wait for the update thread
+        bool expected = false;
+        while (rebuild_in_progress.compare_exchange_strong(
+            expected, true, std::memory_order_relaxed)) {
+            std::this_thread::yield();
+        }
+    }
+
+    table_accessor new_accessor() { return table_accessor{*this}; }
+
+    std::unique_ptr<table_accessor> new_accessor_up()
+    {
+        return std::make_unique<table_accessor>(*this);
+    }
+
+    std::size_t approx_size(table &table) const { return table.size.load(); }
+
+private:
+    void rebuild_table(table &old_table, std::uint32_t report_threshold)
+    {
+        auto new_table = std::make_unique<table>();
+
+        struct copiable_entry {
```

Comment:
nit: copyable?

Reply:
"copiable" is the [canonical](https://en.wiktionary.org/wiki/copiable) form.